adam/learner/affordances.py

```diff
@@ -233,3 +248,97 @@ def _fallback_templates(
                         if isinstance(node, ObjectSemanticNodePerceptionPredicate)
                     }
                 ), 1.0
+
+
+@attrs(slots=True)
+class MappingAffordanceLearner(AbstractAffordanceTemplateLearner):
+
+    concept_to_affordances: MutableMapping[Concept, MutableSet[Concept]] = attrib(
+        validator=deep_mapping(instance_of(Concept), deep_iterable(instance_of(Concept))),
+        factory=dict,
+    )
+    affordance_debug_string_to_concept: MutableMapping[str, Concept] = attrib(
+        validator=deep_mapping(instance_of(str), instance_of(Concept)), factory=dict
+    )
+
+    def _match_template(
+        self,
+        *,
+        concept: Concept,
+        pattern: PerceptionGraphTemplate,
+        perception_graph: PerceptionGraph,
+        confidence: float,
+    ) -> Iterable[Tuple[PerceptionGraphPatternMatch, SemanticNode]]:
+        return immutableset()
+
+    def _new_concept(self, debug_string: str) -> Concept:
+        return AffordanceConcept(debug_string)
+
+    def _learning_step(
+        self,
+        language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> None:
+        # This loop should only execute once in standard cases
+        # As a normal affordance only has one slot filler,
+        # the object the affordance affects.
```

Comment:
This scanned a little weird to me, just suggesting some minor changes. Otherwise looks good, thanks for adding.

```suggestion
        # This loop should only execute once in standard cases
        # because a normal affordance only has one slot filler, namely
        # the object that has the affordance.
```